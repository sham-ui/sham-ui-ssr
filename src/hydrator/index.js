import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
