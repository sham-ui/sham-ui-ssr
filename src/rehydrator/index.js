import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
