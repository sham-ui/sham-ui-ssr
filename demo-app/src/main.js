import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

