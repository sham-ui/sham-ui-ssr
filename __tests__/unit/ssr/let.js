import { compileAsSFC, compile } from 'sham-ui-test-helpers';
import { ssr } from '../../../src/testing';

it( 'should single file component work', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compileAsSFC`
        <template>
            {% let sum = a + b %}
            <ul>
                <li>{{ sum }}</li>
                <li>{{ sum - a }}</li>
                <li>{{ sum - b }}</li>
            </ul>
        </template>
        
        <script>
            export default Component( Template );
        </script>
        `,
        {
            a: 1,
            b: 2
        }
    );
    expect( meta.html ).toBe( '<ul><li>3</li><li>2</li><li>1</li></ul>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should single file component and methods work', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compileAsSFC`
        <template>
            {% let sum = this.sum( a, b ) %}
            <ul>
                <li>{{ sum }}</li>
                <li>{{ sum - a }}</li>
                <li>{{ sum - b }}</li>
            </ul>
        </template>
        
        <script>
            export default Component( Template, function() {
                this.sum = function( a, b ) {
                    return a + b;
                }
            } );
        </script>
        `,
        {
            a: 1,
            b: 2
        }
    );
    expect( meta.html ).toBe( '<ul><li>3</li><li>2</li><li>1</li></ul>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work in for loop', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compileAsSFC`
        <template>
            {% let sum = this.sum( arr ) %}
            <ul>
                {% for item of arr %}
                    <li>{{ sum - item }}</li>
                {% endfor %}
            </ul>
        </template>
        
        <script>
            export default Component( Template, function() {
                this.sum = function( arr ) {
                    return arr.reduce( ( acc, x ) => acc + x, 0 );
                }
            } )
        </script>
        `,
        {
            arr: [ 1, 2, 3 ]
        }
    );
    expect( meta.html ).toBe( '<ul><li>5</li><li>4</li><li>3</li></ul>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with const', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            {% let sum = 42 %}
            {{ sum * 2 }}
        `
    );
    expect( meta.html ).toBe( '84' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should\'t update let vars', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            {% let sum = 42 %}
            {{ sum * 2 }}
        `
    );
    expect( meta.html ).toBe( '84' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should correct build spots order', async() => {
    expect.assertions( 2 );
    const meta = await ssr(
        compile`
            {% let sum = a + b + c %}
            <ul>
                <li>{{ sum }}</li>
                <li>{{ sum - a }}</li>
                <li>{{ a + b }}</li>
            </ul>
        `,
        {
            a: 1,
            b: 2,
            c: 3
        }
    );
    expect( meta.html ).toBe( '<ul><li>6</li><li>5</li><li>3</li></ul>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
