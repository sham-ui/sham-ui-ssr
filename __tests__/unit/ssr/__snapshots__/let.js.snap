// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correct build spots order 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                0,
                0
            ],
            [
                0,
                1
            ],
            [
                0,
                1,
                0
            ],
            [
                0,
                2
            ],
            [
                0,
                2,
                0
            ]
        ]
    }
}",
  "html": "
<ul>
  <li>6</li>
  <li>5</li>
  <li>3</li>
</ul>
",
}
`;

exports[`should single file component and methods work 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                0,
                0
            ],
            [
                0,
                1
            ],
            [
                0,
                1,
                0
            ],
            [
                0,
                2
            ],
            [
                0,
                2,
                0
            ]
        ]
    }
}",
  "html": "
<ul>
  <li>3</li>
  <li>2</li>
  <li>1</li>
</ul>
",
}
`;

exports[`should single file component work 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ],
            [
                0,
                0,
                0
            ],
            [
                0,
                1
            ],
            [
                0,
                1,
                0
            ],
            [
                0,
                2
            ],
            [
                0,
                2,
                0
            ]
        ]
    }
}",
  "html": "
<ul>
  <li>3</li>
  <li>2</li>
  <li>1</li>
</ul>
",
}
`;

exports[`should work in for loop 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    },
    \\"2\\": {
        \\"N\\": [
            1,
            [
                1,
                0
            ]
        ]
    },
    \\"3\\": {
        \\"N\\": [
            2,
            [
                2,
                0
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\",
            \\"2\\",
            \\"3\\"
        ],
        \\"N\\": [
            0
        ]
    }
}",
  "html": "
<ul>
  <li>5</li>
  <li>4</li>
  <li>3</li>
</ul>
",
}
`;

exports[`should work with const 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0
        ]
    }
}",
  "html": "84",
}
`;

exports[`should't update let vars 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0
        ]
    }
}",
  "html": "84",
}
`;
